server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -198,6 +216,77 @@ public DataType<?> partialType() {
         return new StateType(boundSignature.argTypes().getFirst());
     }
 
+    @Nullable
+    @Override
+    public DocValueAggregator<?> getDocValueAggregator(LuceneReferenceResolver referenceResolver,
+                                                       List<Reference> aggregationReferences,
+                                                       DocTableInfo table,
+                                                       List<Literal<?>> optionalParams) {
+        if (aggregationReferences.size() != 1) {
+            return null;
+        }
+        Reference reference = aggregationReferences.get(0);
+        if (!reference.hasDocValues()) {
+            return null;
+        }
+
+        int limit;
+        if(optionalParams.size() == 1) {
+            limit = (int) optionalParams.getFirst().value();
+        } else {
+            limit = DEFAULT_LIMIT;
+        }
+        return getDocValueAggregator(reference, limit);
+    }
+
+    @Nullable
+    private DocValueAggregator<?> getDocValueAggregator(Reference ref, int limit) {
+        if (!ref.hasDocValues()) {
+            return null;
+        }
+        return switch (ref.valueType().id()) {
+            case ByteType.ID,
+                 ShortType.ID,
+                 IntegerType.ID,
+                 LongType.ID,
+                 TimestampType.ID_WITH_TZ,
+                 TimestampType.ID_WITHOUT_TZ,
+                 FloatType.ID,
+                 DoubleType.ID,
+                 GeoPointType.ID ->
+                new SortedNumericDocValueAggregator<>(
+                    ref.storageIdent(),
+                    (ramAccounting, _, _) -> {
+                        int maxMapSize = maxMapSize(limit);
+                        ramAccounting.addBytes(calculateRamUsage(maxMapSize));
+                        var state = new TopKState(new ItemsSketch<>(maxMapSize), limit);
+                        ramAccounting.addBytes(RamUsageEstimator.shallowSizeOf(state));
+                        return state;
+                    },
+                    (values, state) -> {
+                        state.itemsSketch.update(values.nextValue());
+                    }
+                );
+            case IpType.ID,
+                 StringType.ID -> new BinaryDocValueAggregator<>(
+                ref.storageIdent(),
+                (ramAccounting, _, _) -> {
+                    int maxMapSize = maxMapSize(limit);
+                    ramAccounting.addBytes(calculateRamUsage(maxMapSize));
+                    var state = new TopKState(new ItemsSketch<>(maxMapSize), limit);
+                    ramAccounting.addBytes(RamUsageEstimator.shallowSizeOf(state));
+                    return state;
+                },
+                (values, state) -> {
+                    long ord = values.nextOrd();
+                    BytesRef value = values.lookupOrd(ord);
+                    state.itemsSketch.update(value.utf8ToString());
```

Comment:
Doesn't the IP type require additional decoding logic as done in:

https://github.com/crate/crate/blob/c8f7d8313d6e2a5aa252f50187a8bfb7f419389a/server/src/main/java/io/crate/expression/reference/doc/lucene/IpColumnReference.java#L53-L54 

?

Reply:
Thanks for the feedback. I will tackle any other types than string and long as a follow-up.